src/transformers/generation_tf_utils.py

```diff
@@ -2260,6 +2305,511 @@ def sample(
+            [p for p in logits_processor if isinstance(p, TFRepetitionPenaltyLogitsProcessor)]
+        )
+        logits_processor = TFLogitsProcessorList(
+            [p for p in logits_processor if not isinstance(p, TFRepetitionPenaltyLogitsProcessor)]
+        )
+
+        max_length = max_length if max_length is not None else self.config.max_length
+        pad_token_id = pad_token_id if pad_token_id is not None else self.config.pad_token_id
+        eos_token_id = eos_token_id if eos_token_id is not None else self.config.eos_token_id
+        num_return_sequences = (
+            num_return_sequences if num_return_sequences is not None else self.config.num_return_sequences
+        )
+
+        output_attentions = output_attentions if output_attentions is not None else self.config.output_attentions
+        output_hidden_states = (
+            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
+        )
+        output_scores = output_scores if output_scores is not None else self.config.output_scores
+        return_dict_in_generate = (
+            return_dict_in_generate if return_dict_in_generate is not None else self.config.return_dict_in_generate
+        )
+
+        length_penalty = length_penalty if length_penalty is not None else self.config.length_penalty
+        early_stopping = early_stopping if early_stopping is not None else self.config.early_stopping
+
+        use_xla = not tf.executing_eagerly()
+        # GPT2 has a slightly different cache structure, with a different batch axis
+        cache_batch_axis = 1 if "TFGPT2" in str(self) else 0
+
+        # 2. init `attentions`, `hidden_states`, and `scores` tuples
+        scores = [] if (return_dict_in_generate and output_scores) else None
+        decoder_attentions = [] if (return_dict_in_generate and output_attentions) else None
+        cross_attentions = [] if (return_dict_in_generate and output_attentions) else None
+        decoder_hidden_states = [] if (return_dict_in_generate and output_hidden_states) else None
+
+        # 3. init tensors to use for "xla-compileable" generate function
+        batch_size, num_beams, cur_len = input_ids.shape
+
+        # per batch, beam-item holding current token in loop.
+        sequences = tf.TensorArray(
+            element_shape=(batch_size, num_beams),
+            dtype=tf.int32,
+            dynamic_size=False,
+            size=max_length,
+            clear_after_read=False,
+        )
+        running_sequences = tf.TensorArray(
+            element_shape=(batch_size, num_beams),
+            dtype=tf.int32,
+            dynamic_size=False,
+            size=max_length,
+            clear_after_read=False,
+        )
+        intermediary_running_sequences = tf.TensorArray(
+            element_shape=(batch_size, num_beams * 2),
+            dtype=tf.int32,
+            dynamic_size=False,
+            size=max_length,
+            clear_after_read=False,
+        )
```

Comment:
At some point in the loop, we need to write tokens into a structure with `2*num_beams`. There were 3 options:
1. Add an additional `tf.TensorArray` to use as scratchpad (what I did)
2. Use XLA's `dynamic_update_slice`, like FLAX (wanted to avoid, since it is not well documented and can be deprecated out of the blue)
3.  Use gather/scatters (painful to understand)

Reply:
I believe that when compiled with XLA, `tf.TensorArray` writes are quite efficient! It's totally fine to use them when you're building an array in slices like that.

Reply:
As a general comment, though, even though the specific `dynamic_update_slice` function we call is undocumented in the TF code, it is a documented XLA instruction that works identically between JAX and TF, so I think it's okay to use in XLA code. They might move the exact location in TF you have to import it from, but I don't think they'll delete the instruction entirely.

Reply:
Agree!